pkg/workload/workload.go

```diff
@@ -698,42 +698,43 @@ func BaseSSAWorkload(w *kueue.Workload, strict bool) *kueue.Workload {
+// in the specified ClusterQueue and updates the Workload status accordingly.
+//
+// Effects:
+//   - Sets w.Status.Admission to the provided admission.
+//   - Adds or updates a Condition of type kueue.WorkloadQuotaReserved with
+//     Status=True, Reason="QuotaReserved", Message="Quota reserved in ClusterQueue <name>",
+//     and ObservedGeneration set to w.Generation. The message is truncated via
+//     api.TruncateConditionMessage.
+//   - Resets any active "evicted" and "preempted" conditions by invoking
+//     resetActiveCondition for kueue.WorkloadEvicted and kueue.WorkloadPreempted.
+func SetQuotaReservation(workload *kueue.Workload, admission *kueue.Admission, clock clock.Clock) {
+	workload.Status.Admission = admission
+	reason := "QuotaReserved"
+	message := fmt.Sprintf("Quota reserved in ClusterQueue %s", workload.Status.Admission.ClusterQueue)
 	admittedCond := metav1.Condition{
 		Type:               kueue.WorkloadQuotaReserved,
 		Status:             metav1.ConditionTrue,
-		Reason:             "QuotaReserved",
+		Reason:             reason,
 		Message:            api.TruncateConditionMessage(message),
-		ObservedGeneration: w.Generation,
-	}
-	apimeta.SetStatusCondition(&w.Status.Conditions, admittedCond)
-
-	// reset Evicted condition if present.
-	if evictedCond := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadEvicted); evictedCond != nil {
-		evictedCond.Status = metav1.ConditionFalse
-		evictedCond.Reason = "QuotaReserved"
-		evictedCond.Message = preservePreviousMessage(evictedCond.Message)
-		evictedCond.LastTransitionTime = metav1.NewTime(clock.Now())
-	}
-	// reset Preempted condition if present.
-	if preemptedCond := apimeta.FindStatusCondition(w.Status.Conditions, kueue.WorkloadPreempted); preemptedCond != nil {
-		preemptedCond.Status = metav1.ConditionFalse
-		preemptedCond.Reason = "QuotaReserved"
-		preemptedCond.Message = preservePreviousMessage(preemptedCond.Message)
-		preemptedCond.LastTransitionTime = metav1.NewTime(clock.Now())
+		ObservedGeneration: workload.Generation,
 	}
+	apimeta.SetStatusCondition(&workload.Status.Conditions, admittedCond)
+	resetActiveCondition(workload.Status.Conditions, kueue.WorkloadEvicted, reason)
+	resetActiveCondition(workload.Status.Conditions, kueue.WorkloadPreempted, reason)
 }
 
-func preservePreviousMessage(message string) string {
-	if !strings.HasPrefix(message, "Previously: ") {
-		return api.TruncateConditionMessage("Previously: " + message)
+func resetActiveCondition(conditions []metav1.Condition, conditionType, reason string) {
+	condition := apimeta.FindStatusCondition(conditions, conditionType)
+	// Ignore not found or inactive condition.
+	if condition == nil || condition.Status != metav1.ConditionTrue {
+		return
 	}
-
-	return message
+	condition.Status = metav1.ConditionFalse
+	condition.Reason = reason
+	condition.Message = api.TruncateConditionMessage("Previously: " + condition.Message)
+	condition.LastTransitionTime = metav1.Now()
```

Comment:
Please use clock.Now().

Reply:
sorry, forgot to push more resent version that doesn't use clock.  PTAL.